guardian/newsletters-nx
# Dblatcher/launch draft to newsletter

## What does this change?

Adds functions to check if a draft has the necessary data to be launched and to set default values for legacy fields not exposed in the wizard.

Allows users to use the launchNewsletter wizard to  "launch" a completed `DraftNewsletterData` into a `NewsletterData`, so it will appear on the API response.

Adds a link to the launch wizard on the draft details pages, if the draft is complete.

## How to test
Launch a 'pre-completed draft', see wizard refuse to launch and imcomplete draft:
 - `npm run dev`
 - Go to http://localhost:4200/demo/launch-newsletter/5770 to open the launch wizard for a fully complete draft. Going through the steps will launch the newsletter, adding it the list at http://localhost:4200/newsletters
 - http://localhost:4200/demo/launch-newsletter/7001 will open the wizard for an incomplete draft. The wizard will not allow you to launch it and include a list of the fields with errors at the second step.

Create and launch a new newsletter:
 - `npm run dev`
 - open the "create draft" wizard and answer all the questions up to "Complete Data Collection"
 - go to the drafts page (http://localhost:4200/drafts) and open the details view for your new newsletter (by clicking the list ID in the first column of the table)
 - the launch🚀  button should be on the details page - press it to access the launch wizard



## How can we measure success?

Users able to start create a new newsletter in the UI so it will be published by the API.

## Have we considered potential risks?

In once sense, we are 'deleting' the draft, but all of the data it contained is used in the newsletter - the operation is structured as creating the newsletter and then deleting the draft, but it could also be thought of as a 'move' operation.

## Images

|launch button| lauched step | not ready step|
|---|---|---|
|<img width="868" alt="Screenshot 2023-04-19 at 16 39 41" src="https://user-images.githubusercontent.com/30567854/233133636-1acb552b-d9e4-4104-9a3d-4e0bdfad4be3.png">|<img width="898" alt="Screenshot 2023-04-19 at 16 40 10" src="https://user-images.githubusercontent.com/30567854/233133694-e88dd26f-51b2-4d5a-a7a6-dbddaae01427.png">|<img width="898" alt="Screenshot 2023-04-19 at 16 40 49" src="https://user-images.githubusercontent.com/30567854/233134364-a568e07c-7908-4f92-928b-ae584d29d2fd.png">|
